open_schools_platform/organization_management/circles/services.py

```diff
@@ -44,6 +48,43 @@ def create_circle(name: str, organization: Organization, description: str, capac
     return circle
 
 
+class CircleQueryHandler(BaseQueryHandler):
+    allowed_statuses = [Query.Status.ACCEPTED, Query.Status.SENT, Query.Status.CANCELED, Query.Status.DECLINED]
+
+    @staticmethod
+    def query_handler(query: Query, new_status: str, user: User):
+        BaseQueryHandler.query_handler_checks(CircleQueryHandler, query, new_status, user)
+
+        from open_schools_platform.parent_management.families.models import Family
+
+        if type(query.recipient) is not Family:
+            raise ValidationError(detail="The recipient must be a Family if the sender is a circle")
+
+        circle_access = user.has_perm('circles.circle_access', query.sender)
+        family_access = user.has_perm('families.family_access', query.recipient)
+
+        if not family_access:
+            if new_status != Query.Status.CANCELED:
+                raise NotAcceptable("Circle can only set canceled status")
+        elif not circle_access:
+            if query.status != Query.Status.SENT:
+                raise NotAcceptable("Сan no longer change the query")
+            if new_status == Query.Status.CANCELED:
+                raise NotAcceptable("User cannot cancel query, he can only decline or accept it")
```

Comment:
what if we haven't access to family and circle both?
should we raise PermissionDenied?

Reply:
I dont think that such a situation is possible when someone who is neither the recipient nor the sender nor the admin can somehow access the information about query and change its status.